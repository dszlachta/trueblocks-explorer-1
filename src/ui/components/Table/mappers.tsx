import React, { ReactElement } from 'react';
import { Tag } from 'antd';
import { option as Option } from 'fp-ts';
import { pipe } from 'fp-ts/function';

export function renderFlag(flag?: boolean): ReactElement {
  return pipe(
    flag,
    Option.fromPredicate((flagOrUndefined) => flagOrUndefined !== undefined),
    Option.fold(
      () => (<></>),
      (someFlag) => (
        <Tag color={someFlag ? 'green' : 'volcano'}>
          {someFlag ? 'Yes' : 'No'}
        </Tag>
      ),
    ),
  );
}

export type TagClickHandler = (rawTag: string) => void;
export function renderTagsWithClickHandler(
  onClick: TagClickHandler,
) {
  return function renderTags(tagsString?: string) {
    return pipe(
      tagsString,
      Option.fromPredicate<string | undefined>(Boolean),
      Option.fold(
        () => (<></>),
        (someTagsString) => (
          <>
            {someTagsString!.split(':').map((tag) => (
              <Tag onClick={() => onClick(someTagsString!)} key={tag}>
                {tag}
              </Tag>
            ))}
          </>
        ),
      ),
    );
  };
}

export function renderActionsAsColumn<Item>(
  columnWidth: number,
) {
  return function getActionsColumn(
    getActions: (record: Item) => ReactElement,
  ) {
    return {
      fixed: 'right' as const,
      width: columnWidth,
      render(text: string, record: Item) {
        return getActions(record);
      },
    };
  };
}
