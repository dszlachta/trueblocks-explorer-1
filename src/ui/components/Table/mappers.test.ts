import renderer from 'react-test-renderer';

import * as Mappers from './mappers';

describe('Table/mappers module', () => {
  test('renderFlag returns correct components', () => {
    const { renderFlag } = Mappers;

    const treeWhenNoFlag = renderer
      .create(
        renderFlag(undefined),
      ).toJSON();

    const treeWhenFlagTrue = renderer
      .create(
        renderFlag(true),
      ).toJSON();

    const treeWhenFlagFalse = renderer
      .create(
        renderFlag(false),
      ).toJSON();

    expect(treeWhenNoFlag).toMatchSnapshot();
    expect(treeWhenFlagTrue).toMatchSnapshot();
    expect(treeWhenFlagFalse).toMatchSnapshot();
  });

  test('renderTagsWithClickHandler returns correct components', () => {
    const renderTags = Mappers.renderTagsWithClickHandler(() => { });
    const tags = '31-Gitcoin Grants:Friends';

    const treeWhenNoTags = renderer
      .create(
        renderTags(undefined),
      ).toJSON();

    const treeWhenTags = renderer
      .create(
        renderTags(tags),
      ).toJSON();

    expect(treeWhenNoTags).toMatchSnapshot();
    expect(treeWhenTags).toMatchSnapshot();
  });

  test('renderActionsAsColumn uses getActions function', () => {
    const getActions = jest.fn();
    const expectedColumnWidth = 100;
    const record = { a: 1 };

    const getColumn = Mappers.renderActionsAsColumn(expectedColumnWidth);
    const result = getColumn(getActions);
    result.render('text', record);

    expect(result.width).toBe(expectedColumnWidth);
    expect(getActions).toHaveBeenCalledWith(record);
  });
});
