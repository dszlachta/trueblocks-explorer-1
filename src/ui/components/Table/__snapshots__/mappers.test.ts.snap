// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table/mappers module renderFlag returns correct components 1`] = `null`;

exports[`Table/mappers module renderFlag returns correct components 2`] = `
<span
  className="ant-tag ant-tag-green"
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  Yes
</span>
`;

exports[`Table/mappers module renderFlag returns correct components 3`] = `
<span
  className="ant-tag ant-tag-volcano"
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  No
</span>
`;

exports[`Table/mappers module renderTagsWithClickHandler returns correct components 1`] = `null`;

exports[`Table/mappers module renderTagsWithClickHandler returns correct components 2`] = `
Array [
  <span
    className="ant-tag"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    31-Gitcoin Grants
  </span>,
  <span
    className="ant-tag"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    Friends
  </span>,
]
`;
